sw/device/lib/base/bitfield.h

```diff
@@ -13,58 +14,101 @@ extern "C" {
 #endif  // __cplusplus
 
 /**
- * Masked field offset for 32-bit bitfields, with optional value.
+ * A field of a 32-bit bitfield.
  *
- * `index` represents a shift in bits starting from the 0 bit of a given 32-bit
- * bitfield. It is used to zero the memory inside this bitfield by applying an
- * inverted `mask` at the `index` offset. `value` is then capped by applying
- * `mask` and is shifted to the `index` bits of the bitfield.
+ * The following field definition: `{ .mask = 0xb11, .index = 12 }`
  *
- * Example:
- * ```
- * 32-bit bitfield = 0b11111111'11111111'11111111'11111111
- * set_field32( mask = 0b11111111, index = 16, value = 0b01111110 (126) )
- * result 32-bit bitfield = 0b11111111'01111110'11111111'11111111
- * ```
+ * Denotes the X-marked bits in the following 32-bit bitfield:
+ *
+ *     field:  0b--------'--------'--XX----'--------
+ *     index:   31                                 0
+ *
+ * Restrictions: The index plus the width of the mask must not be greater than
+ * 31.
  */
 typedef struct bitfield_field32 {
-  uint32_t mask;  /**< The field mask. */
-  uint32_t index; /**< The field position within a bitfield. */
-  uint32_t
-      value; /**< The field value to be written, if using a `set` function. */
+  /** The field mask. Usually all ones. */
+  uint32_t mask;
+  /** The field position in the bitfield, counting from the zero-bit. */
+  uint32_t index;
 } bitfield_field32_t;
 
 /**
- * Gets `field` in `bitfield`.
+ * Reads a value from `field` in `bitfield`.
  *
- * This function uses the `field` parameter to get a value at the given
- * offset (and mask) in `bitfield`. Only `mask` and `index` are used;
- * `value` is ignored.
+ * This function uses the `field` parameter to read the value from `bitfield`.
+ * The resulting value will be unshifted and zero-extended so the field's
+ * zero-bit is in the return value's zero-bit.
  *
  * @param bitfield Bitfield to get the field from.
  * @param field Field to read out from.
- * @return The read value, zero-extended to 32 bits.
+ * @return Zero-extended `field` from `bitfield`.
  */
-inline uint32_t bitfield_get_field32(uint32_t bitfield,
-                                     bitfield_field32_t field) {
+inline uint32_t bitfield_field32_read(uint32_t bitfield,
```

Comment:
A thought: maybe we should mark these as warn_unused_result. Especially with `write`, it's not always clear that the function is stateless and something needs to be done with the result.

Reply:
That raises a question about how we want to use `warn_unused_result`:
- We could use it to ensure errors are always handled (current praxis for dif_*_result_t); and/or
- We could use it more widely to ensure we avoid making mistakes (ie, annotating here).

I do understand the use of the latter, though I worry about how noisy this might turn out.

@alphan @silvestrst thoughts?

Reply:
> Especially with write, it's not always clear that the function is stateless and something needs to be done with the result.

Good point!

> We could use it to ensure errors are always handled (current praxis for dif_*_result_t); and/or
We could use it more widely to ensure we avoid making mistakes (ie, annotating here).

I like the idea.

> I do understand the use of the latter, though I worry about how noisy this might turn out.

I think it works great for this particular use-case, it could be nice to force these warnings to errors (maybe `-Werror=`?). Could you remind me how our build system work? I know we have a set of compiler flags in root `meson.build`, can these be overridden for individual libraries, or do we use the same set of flags for the whole build?

Reply:
We have `-Werror` enabled in builds, for all code. We can turn this off for individual libraries, but I think that we want to avoid that.

Another way of stating the noisyness issue is: how many errors from enabling it here (the second option) would lead to build warnings/errors where there was not a logic error? Thinking further about it, I think the answer is "none", because these are pure functions that don't have side-effects, so if you don't use the result, why call the function at all?

Reply:
> Thinking further about it, I think the answer is "none", because these are pure functions that don't have side-effects, so if you don't use the result, why call the function at all

Yup, that is my understanding as well.

Reply:
FWIW, the only reason I bring this up is because Rust does something similar: https://doc.rust-lang.org/std/primitive.u32.html#method.pow

`u32::pow` is marked `#[must_use]` because it's a pure function but it's not not obvious from reading. Many times I've expected `set_field32` to take a pointer and mutate through it.

Maybe I'm just *too* worried. If you think this is super noisy and won't save us from making mistakes, I'm fine with accepting that this isn't the place.

Reply:
I've added this attribute to all the functions, and we don't get warnings when building software using this branch, which is nice. It points to "this won't be noisy here".

I'm not sure we've resolved the macro discussion yet, the patch I have uploaded uses one for the attribute, and the formatting is good, but that's not how we've approached this before. I'll update this with whatever #2604 decides, rather than having two discussions.

Reply:
#2604 resolved to say macros for attributes are allowed. In that PR we introduced a macro-specific header for that attribute in DIFs - in this PR I'm happy to proceed with having defines in the top of the header.

I missed using `__attribute__((warn_unused_result))` in mmio.h until today's updates - that is now resolved, using a macro.